core/src/test/java/com/backblaze/b2/json/B2JsonInferredParametersTest.java

```diff
@@ -0,0 +1,2858 @@
+/*
+ * Copyright 2018, Backblaze Inc. All Rights Reserved.
+ * License https://www.backblaze.com/using_b2_code.html
+ */
+
+package com.backblaze.b2.json;
+
+import com.backblaze.b2.util.B2BaseTest;
+import com.backblaze.b2.util.B2Preconditions;
+import com.backblaze.b2.util.B2StringUtil;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.lang.reflect.Type;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.regex.Pattern;
+
+import static org.junit.Assert.*;
+
+/**
+ * Unit tests for B2Json relying on javac's '-parameters' option instead of {@link B2Json.constructor#params}.
+ */
+@SuppressWarnings({
+        "unused",  // A lot of the test classes have things that aren't used, but we don't care.
+        "WeakerAccess"  // A lot of the test classes could have weaker access, but we don't care.
+})
+public class B2JsonInferredParametersTest extends B2BaseTest {
+
+    @Rule
+    public ExpectedException thrown  = ExpectedException.none();
+
+    private static final class Container {
+
+        @B2Json.required
+        public final int a;
+
+        @B2Json.optional
+        public final String b;
+
+        @B2Json.ignored
+        public int c;
+
+        @B2Json.constructor
+        public Container(int a, String b) {
+            this.a = a;
+            this.b = b;
+            this.c = 5;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (!(o instanceof Container)) {
+                return false;
+            }
+            Container other = (Container) o;
+            return a == other.a && (b == null ? other.b == null : b.equals(other.b));
+        }
+    }
+
+    private static final B2Json b2Json = B2Json.get();
+
+    @Test
+    public void testBoolean() throws IOException, B2JsonException {
```

Comment:
Do these tests (other than those that use your `Container` nested class and the others that don't use `@param`) somehow test something new? I'm not clear how they have anything to do with your newly introduced functionality.

Reply:
Oh I see. So they test that your new functionality didn't indirectly break something else?

Reply:
Yes. They test that all the existing (tested) features play nicely with this new functionality. Think of it as way to test for harmful 'drug interactions' with no medical knowledge. I'm new to this library, so don't have an intuition for which features are more likely to interact poorly with this. Instead, just test them all.

Now that we're shown there to be no poor interactions, having just a few focused unit tests will keep this more maintainable.

Reply:
Done